File: 8-blocks_lambdas.rb
# Ruby block 係一等公民，method 可以直接收 &block。
def repeat(times, &block)
  times.times(&block)
end

repeat(3) do |index|
  puts "Ruby block ##{index + 1}"
end

puts '---'

# lambda 可以保存外層 number，之後再 call。
def build_multiplier(number)
  lambda do |value|
    value * number
  end
end

double = build_multiplier(2)
triple = build_multiplier(3)

puts double.call(10)
puts triple.call(10)

puts '---'

numbers = [1, 2, 3, 4]
squares = numbers.map { |number| number * number }
evens = numbers.select { |number| number.even? }

p squares
p evens

File: 8-blocks_lambdas.py
# Python 通常將 function 或 lambda 當 callback 傳入另一個 function。
def repeat(times, action):
    for index in range(times):
        action(index)


repeat(3, lambda index: print(f"Python lambda #{index + 1}"))

print("---")


# nested function 可以保存外層 number，形成 closure。
def build_multiplier(number):
    def multiplier(value):
        return value * number

    return multiplier


double = build_multiplier(2)
triple = build_multiplier(3)

print(double(10))
print(triple(10))

print("---")


numbers = [1, 2, 3, 4]
squares = list(map(lambda number: number * number, numbers))
evens = list(filter(lambda number: number % 2 == 0, numbers))

print(squares)
print(evens)
